qiskit/algorithms/evolvers/pvqd/pvqd.py

```diff
@@ -0,0 +1,390 @@
+
+                This is hardcoded here since the gradient framework does not support computing
+                gradients for overlaps.
+
+                Args:
+                    displacement: The parameters for the ansatz.
+
+                Returns:
+                    The gradient.
+                """
+                # construct lists where each element is shifted by plus (or minus) pi/2
+                dim = displacement.size
+                plus_shifts = (displacement + np.pi / 2 * np.identity(dim)).tolist()
+                minus_shifts = (displacement - np.pi / 2 * np.identity(dim)).tolist()
+
+                evaluated = evaluate_loss(plus_shifts + minus_shifts)
+
+                gradient = (evaluated[:dim] - evaluated[dim:]) / 2
+
+                return gradient
+
+        else:
+            evaluate_gradient = None
+
+        return evaluate_loss, evaluate_gradient
+
+    def evolve(self, evolution_problem: EvolutionProblem) -> EvolutionResult:
+        """
+        Args:
+            evolution_problem: The evolution problem containing the hamiltonian, total evolution
+                time and observables to evaluate. Note that :class:`~.PVQD` currently does not support
+                hamiltonians of type :class:`~.MatrixOp`.
+
+        Returns:
+            A result object containing the evolution information and evaluated observables.
+
+        Raises:
+            ValueError: If the evolution time is not positive or the timestep is too small.
+            NotImplementedError: If the evolution problem contains an initial state.
+        """
+        self._validate_setup()
+
+        time = evolution_problem.time
+        observables = evolution_problem.aux_operators
+        hamiltonian = evolution_problem.hamiltonian
+
+        if not 0 < self.timestep <= time:
+            raise ValueError(
+                f"The time step ({self.timestep}) must be larger than 0 and smaller equal "
+                f"the evolution time ({time})."
+            )
+
+        if evolution_problem.initial_state is not None:
+            raise NotImplementedError(
+                "Setting an initial state for the evolution is not yet supported for PVQD."
+            )
+
+        # get the function to evaluate the observables for a given set of ansatz parameters
+        if observables is not None:
+            evaluate_observables = _get_observable_evaluator(
+                self.ansatz, observables, self.expectation, self._sampler
+            )
+            observable_values = [evaluate_observables(self.initial_parameters)]
+
+        fidelities = [1]
+        times = [0]
+        parameters = [self.initial_parameters]
+
+        current_time = 0
+        initial_guess = self.initial_guess
+
+        while current_time < time:
+            # perform VQE to find the next parameters
+            next_parameters, fidelity = self.step(
+                hamiltonian, self.ansatz, parameters[-1], self.timestep, initial_guess
+            )
```

Comment:
Given that timestep must be > 0 and <= time what is the expectation of the behavior of this loop. If we step along at timestep from 0 we may exceed the time. Eg my timestep is 0.5 and my time is 12.6 I would get to 12.5 and current time would be less than time and I will go to 13 and pass back a result for time=13 rather than 12.6?

Reply:
Yes, the expectation would be that you return the smallest time above or equal to the final time 😄 We could also change such that you give a final time and a number of timesteps.

Reply:
Giving back an answer to a slightly different problem - do we want to do that. If I want the result at t=12.6 should that not be the case? (The result does not have field to state the time we actually evolved it until, versus what your problem asked for) As you say the number of steps could be a solution. Another, if its possible, would be to make the last time step smaller than the supplied step so that hits the target time rather than going beyond.

Reply:
Resolved by switching to `num_timesteps`